### PR TITLE
Update MMCVRayEstimator return value, support mmcv batch_processor 

### PR DESCRIPTION
**Pull request by PatrickkZ**
## Description

- Fix the wrong calculation of `num_samples` in the return value of MMCVRayEstimator fit/run method
- Add support for mmcv `batch_processor`
- update mmcv unit test 

### 1. Why the change?

- The current calculation of  total `num_samples` in the return value of MMCVRayEstimator fit/run method is wrong, this pr fix this problem.
- In some MMCV models, they use `batch_processor` to process the training, this is an obsolete way. this pr add support for `batch_processor` .

### 2. User API changes

User can create a mmcv runner with batch_processor, and use this runner to create a `MMCVRayEstimator`.
```python
class Model(nn.Module):

    def __init__(self):
        super().__init__()
        self.fc1 = nn.Linear(50, 50)
        self.relu1 = nn.ReLU()
        self.dout = nn.Dropout(0.2)
        self.fc2 = nn.Linear(50, 100)
        self.prelu = nn.PReLU(1)
        self.out = nn.Linear(100, 1)
        self.out_act = nn.Sigmoid()
        self.loss_fn = nn.BCELoss()

    def forward(self, input_, labels):
        a1 = self.fc1(input_)
        h1 = self.relu1(a1)
        dout = self.dout(h1)
        a2 = self.fc2(dout)
        h2 = self.prelu(a2)
        a3 = self.out(h2)
        y = self.out_act(a3)
        loss = self.loss_fn(y, labels)
        return loss


# define a batch_processor
def batch_processor(model, data, train_mode, **kwargs):
    features, labels = data
    loss = model(features, labels)
    log_vars = dict()
    log_vars["var1"] = 0.0
    return {'loss': loss, 'log_vars': log_vars, "num_samples": features.size(0)}


def runner_creator(config):
    model = Model()

    optimizer = optim.SGD(model.parameters(), lr=0.001, momentum=0.9)
    logger = get_logger('mmcv')
    # create a runner with batch_processor
    runner = EpochBasedRunner(
        model,
        batch_processor=batch_processor,
        optimizer=optimizer,
        work_dir='./work_dir',
        logger=logger,
        max_epochs=MAX_EPOCH)

    # learning rate scheduler config
    lr_config = dict(policy='step', step=[2, 3])
    # configuration of optimizer
    optimizer_config = dict(grad_clip=None)
    # save log periodically and multiple hooks can be used simultaneously
    log_config = dict(interval=4, hooks=[dict(type='TextLoggerHook')])
    # register hooks to runner and those hooks will be invoked automatically
    runner.register_training_hooks(
        lr_config=lr_config,
        optimizer_config=optimizer_config,
        log_config=log_config)

    return runner
```

use the runner to create a `MMCVRayEstimator`
```python
estimator = MMCVRayEstimator(
        mmcv_runner_creator=runner_creator,
        config={}
    )
```

### 3. Summary of the change 

- Fix the wrong calculation problem of the MMCVRayEstimator fit/run method return value.
- Add support for mmcv batch_processor.
- update ut

### 4. How to test?
- [ ] N/A
- [x] Unit test
- [ ] Application test
- [ ] Document test
- [ ] ...

### CODE REVIEW

File: python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py [hunk 1]
```diff
@@ -165,13 +173,10 @@ def run_iter(self, data_batch: Any, train_mode: bool, **kwargs) -> None:
...
                         copy[k] = v
-                metric_meters.update(copy, n=len(data_batch))
-        else:
-            outputs = self.model.val_step(data_batch, self.optimizer, **kwargs)
-        if not isinstance(outputs, dict):
-            invalidInputError(False,
-                              '"batch_processor()" or "model.train_step()" '
-                              'and "model.val_step()" must return a dict')
+                copy.pop("log_vars", None)
+                copy.pop("num_samples", None)
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py:177]: name `copy` may raise ambiguity with `import copy`,  change to another name.

File: python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py [hunk 2]
```diff
@@ -20,7 +20,7 @@
 from mmcv.runner import EpochBasedRunner
 from mmcv.runner.utils import get_host_info
 from mmcv.parallel.distributed import MMDistributedDataParallel
-from bigdl.orca.learn.pytorch.utils import AverageMeterCollection
+from bigdl.orca.learn.pytorch.utils import AverageMeterCollection, get_batchsize
 from bigdl.dllib.utils.log4Error import *
 from bigdl.orca.learn.pytorch.experimential.core.base_ray_runner import BaseRayRunner
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py:25]: Redirect Trainer to Trainable in BaseRayRunner

**Comment by leonardozcm** [comment]
Other suggestions:
1. Add comments upon `self.model = MMDistributedDataParallel(**)` to explain why use MMDist
2. Comments in `run()`: keep **the** comsistent **format** with
3.  line 47 in mmcv estimator, has already been included in `self.setup` right? and may we think about a better design of self.setup, for now self.setup(params, **self.backend, self.runner_cls**, workers_per_node) seems quite redundant. 
4. line 24 in ray runner `from bigdl.dllib.utils.log4Error import *`, do not import all for this is bad for type hint, make it more clear  `from bigdl.dllib.utils.log4Error import invalidInputError`
5. kwargs["_metric_meters"]
6. We have no way to get train status if user pass a batch_processor?
7. It looks like mmcv.log_buffer has something in common with our AverageMeterCollection, can we merge them?

File: python/orca/test/bigdl/orca/learn/ray/experimental/test_mmcv_estimator_ray_backend.py [hunk 3]
```diff
@@ -110,32 +172,64 @@ def __len__(self):
...
+        estimator = get_estimator(runner_creator)
+        self.call_fit(estimator)
+
+    def test_fit_with_batch_processor(self):
+        estimator = get_estimator(runner_creator_with_batch_processor)
+        self.call_fit(estimator)
+
+    def test_run_with_train_step(self):
+        estimator = get_estimator(runner_creator)
+        self.call_run(estimator)
```
Comment (leonardozcm) [inline python/orca/test/bigdl/orca/learn/ray/experimental/test_mmcv_estimator_ray_backend.py:227]: No need to test on run, just `fit `is ok i think

File: python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py [hunk 4]
```diff
@@ -162,23 +162,25 @@ def run_iter(self, data_batch: Any, train_mode: bool, **kwargs) -> None:
...
         else:
             outputs = self.model.val_step(data_batch, self.optimizer, **kwargs)
         if not isinstance(outputs, dict):
             invalidInputError(False,
                               '"batch_processor()" or "model.train_step()" '
                               'and "model.val_step()" must return a dict')
+
         if 'log_vars' in outputs:
-            self.log_buffer.update(outputs['log_vars'], outputs['num_samples'])
+            if 'loss' in outputs['log_vars']:
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py:173]: make some simple comments here like `ensure key loss is in log_buffer, and driver can fetch loss infos from workers' log_buff`

File: python/orca/test/bigdl/orca/learn/ray/experimental/test_mmcv_estimator_ray_backend.py [hunk 5]
```diff
@@ -110,32 +172,52 @@ def __len__(self):
...
 
 
 class TestMMCVRayEstimator(unittest.TestCase):
 
-    def test_fit(self):
-        estimator = get_estimator()
-        epoch_stats = estimator.fit([train_dataloader_creator], [('train', 1)])
+    def test_run_with_train_step(self):
+        estimator = get_estimator(runner_creator)
+        epoch_stats = estimator.run([train_dataloader_creator], [('train', 1)])
```
Comment (leonardozcm) [inline python/orca/test/bigdl/orca/learn/ray/experimental/test_mmcv_estimator_ray_backend.py:193]: train epochs  > 3

File: python/orca/test/bigdl/orca/learn/ray/experimental/test_mmcv_estimator_ray_backend.py [hunk 6]
```diff
@@ -63,6 +63,39 @@ def train_step(self, data, optimizer, **kwargs):
...
+        y = self.out_act(a3)
+        loss = self.loss_fn(y, labels)
+        return loss
+
+
+def batch_processor(model, data, train_mode, **kwargs):
+    features, labels = data
+    loss = model(features, labels)
+    log_vars = dict()
+    log_vars["var1"] = 1.0
```
Comment (leonardozcm) [inline python/orca/test/bigdl/orca/learn/ray/experimental/test_mmcv_estimator_ray_backend.py:95]: how do you test with `var1`, we need to make sure your log _buff design is working.

File: python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py [hunk 7]
```diff
@@ -138,8 +139,7 @@ def train(self, data_loader, **kwargs):
         self.data_loader = data_loader
         self._max_iters = self._max_epochs * len(self.data_loader)
         self.call_hook('before_train_epoch')
-        metric_meters = AverageMeterCollection()
-        kwargs["metric_meters"] = metric_meters
+        self.log_buffer.clear()
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py:142]: It seems like LoggerHook is doing this: https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/hooks/logger/base.py#L145
it this necessary to doing this? for maybe there are some case they want to calculate the overall average log?

File: python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py [hunk 8]
```diff
@@ -150,7 +150,7 @@ def train(self, data_loader, **kwargs):
             del self.data_batch
             self._iter += 1
 
-        stats = metric_meters.summary()
+        stats = self._get_epoch_stats()
```
Comment (leonardozcm) [inline python/orca/src/bigdl/orca/learn/pytorch/experimential/mmcv/mmcv_ray_runner.py:153]: behavior comment here
